### ISSUE TITLE
Generic derivation of schemas

### ISSUE THREAD

**Issue by vil1** [scalaz/scalaz-schema#48]
Building upon #47, it seems rather obvious that we should be able to automatically derive the schema of any ADT at compile time (as a matter of fact, [this gist](https://gist.github.com/vil1/7f28126db61ee9d6d0a9407da1b042fa) achieves just that, using shapeless).

Like to #47, this should be provided in the `generic` (sbt) module, with no impact on the existing API for constructing schemas.

Also, if #34 gets merged before this issue is solved, words must be added in the documentations to stress out the fact that such fully automatic derivation should be used with caution, for it somehow defeats the purpose of having schemas as runtime values (in scenarios where schema evolution is a needed feature, it OK to fully derive schemas at compile time, as long as the application stores the derived schema somewhere at runtime).

**Comment by sasiharan** [comment]
@vil1,  Had a discussion with @GrafBlutwurst about the gist code you shared in the issue for understanding. I will modify this gist code to align with latest schema encoding and raise a PR to take it forward.

**Comment by vil1** [comment]
@sasiharan that's great news!

FTR, the mentioned gist is [here](https://gist.github.com/vil1/7f28126db61ee9d6d0a9407da1b042fa)

Please note that it might be easier to first work on #47 and then use that work to solve this issue (that is, the stuff around the `Constructor` typeclass in the gist).

Also, I'd prefer we use scalaz-deriving for this instead of shapeless for our generic derivation mechanisms. I'm not yet knowledgeable with it to be able to give you any guidance, but I intend to learn more about it soon (and I can also bring in people who know it well).

**Comment by sasiharan** [comment]
Agreed that #44 and #47 needs to be addressed before working on this issue. I will also go through scalaz-deriving and see how we can achieve auto derivation of Isos and Schemas.

**Comment by vil1** [comment]
Don't worry too much about #44 though, it might take a while to be solved. And if it ever has an impact on your work, I'll take care of it as much as possible.
